### DISCUSSION TITLE
Discord Support

### DISCUSSION THREAD

**Discussion by johnaqu1no** [rutgerkok/WorldGeneratorApi#40] [Q&A] [answered]
Possibly add Discord Support for us to join and ask for support!
I have a few questions that I need answered!

<3

**Comment by johnaqu1no** [comment]
But if you can't, my 2 questions are:
- How do I get the chunk using #getZoomedOutBiome();?
Here is what I currently have, I am not sure if this is remotely correct:
```java
        int actualX = x * 4;
        int actualY = y * 4;
        int actualZ = z * 4;

        Location loc = new Location(world, actualX, actualY, actualZ);
        Utils.broadcast(loc.toString());
        Chunk chunk = world.getChunkAt(loc);
```

- How do I change the biome of a chunk? Not just individual blocks, just the whole chunk (which changes the blocks biome at once)

**Comment by rutgerkok** [comment]
I don't know if the community is large enough for a Discord.

For your first question: I'm not sure what you are coding. Are you implementing a custom biome generator? Then you cannot access chunks, as those don't exist yet when generating terrain.

For the second question: if the chunk already exists, just use the Bukkit API to change biomes, see https://bukkit.org/threads/change-biome.73734/ . You don't need WorldGeneratorAPI for that. If you want to set the biome for new chunks, create your own biome generator.

Have a look at this page: https://github.com/rutgerkok/WorldGeneratorApi/wiki/Biome-generators . It contains examples on how to modify the vanilla biome generator, as well as to how you could create a custom one.

**Comment by johnaqu1no** [comment]
Even though the community is small, new people like I have direct contact with you and the contributors rather than just waiting with a simple pull request.

1. I am basically randomizing the chunks with a random biome. Which is why giving a block a random biome is chaotic, I at least want to make the terrain noticeable with their corresponding chunk.
![image](https://user-images.githubusercontent.com/41652412/118200099-462d2080-b409-11eb-8480-dca3265ecf5e.png)

2. That's the problem I have. I want the biome to have its corresponding blocks that populate that chunk. For instance: `SOUL_SAND_VALLEY` would contain: `BONE_BLOCK, SOUL_SOIL, SOUL_SAND, etc`
Overwriting an existing chunk with a biome won't regenerate that chunk (and even using world edit's //regen command, it wont regenerate to the new biome due to the ChunkData. 

Using the ChunkGenerator class works, although we would need to use the vanilla populator which is not supplied in the spigot api. Which is why I am using WorldGeneratorAPI

**Comment by rutgerkok** [comment, marked as answer]
Ok, I've set up a small discussion forum. But please don't expect me to provide too much support, I've got many other things to do. I've made the code public in the hope that it will be useful to others, not to assist everyone in building their own terrain generator. (That was the goal back when I was maintaining [TerrainControl](https://www.spigotmc.org/resources/terraincontrol.2214/), but that costed waaaaay too much time.)

If you assign biomes using a BiomeGenerator, then the vanilla terrain generator should automatically generate the correct terrain for that biome. If you use WorldEdit and //regen, then the terrain should still correctly generate.

From https://github.com/rutgerkok/WorldGeneratorApi/issues/38 , it looks like you're already using a BiomeGenerator, so I'm not sure what is going on here. Something like this should work as a random biome generator. (It still consistently returns the same biome for a given (x,y,z), which is required, otherwise the biome can change mid-generation.)

```java
    private static class YourBiomeGenerator implements BiomeGenerator {

        private final ImmutableList<Biome> allBiomes = ImmutableList
                .copyOf(BiomeGenerator.VANILLA_OVERWORLD_STRUCTURE_BIOMES);
        private final ImmutableSet<Biome> allBiomesAsSet = ImmutableSet.copyOf(allBiomes);
        
        private final YourConfigClass config;

        private YourBiomeGenerator (YourConfigClass config) {
            this.config = config;
        }

        @Override
        public ImmutableSet<Biome> getStructureBiomes() {
            return allBiomesAsSet;
        }

        @Override
        public Biome getZoomedOutBiome(int x, int y, int z) {
            long seed = config.biomeGeneratorSeed;
            Random random = new Random(seed ^ x ^ z);
            return allBiomes.get(random.nextInt(allBiomes.size()));
        }
    }
```
